File: Lista Enlazada/Lista.cpp
#include <iostream>
#include <cstdlib>
using namespace std;

struct node {
    int data;
   struct node *next;
};

struct node *head;

void insertbeg();
void insertend();
void randominsert();
void deletebeg();
void deleteend();
void randomdelete();
void display();
void search();

int main() {
    int choice = 0;
    while (choice != 9) {
        cout << "\n*********Main Menu*********\n";
        cout << "1. Insert at beginning\n";
        cout << "2. Insert at end\n";
        cout << "3. Insert at random location\n";
        cout << "4. Delete from beginning\n";
        cout << "5. Delete from end\n";
        cout << "6. Delete from random location\n";
        cout << "7. Display\n";
        cout << "8. Search\n";
        cout<<"9. Exit\n";
        cout << "Enter your choice: ";
        cin >> choice;
        switch (choice) {
            case 1:
                insertbeg();
                break;
            case 2:
                insertend();
                break;
            case 3:
                randominsert();
                break;
            case 4:
                deletebeg();
                break;
            case 5:
                deleteend();
                break;
            case 6:
                randomdelete();
                break;
            case 7:
                display();
                break;
            case 8:
                search();
                break;
            case 9:
                exit(0);
            default:
                cout << "Invalid choice! Please try again.\n";
        }
    }
    return 0;
}

void insertbeg() {
    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    int item;
    if (newnode == NULL) {
        cout << "OVERFLOW\n";
        
    }else{
         cout << "Ingresa un elemento: ";
    cin >> item;
    newnode->data = item;
    newnode ->next = head;
    head = newnode;
    cout << "Elemento insertado\n";
    }
   
}

void insertend() {
    struct node *newnode,*temp;
    newnode = (struct node *)malloc(sizeof(struct node));
    int item;
    if (newnode == NULL) {
        cout << "OVERFLOW\n";
        
    }else{
         cout << "Ingresa un elemento: ";
    cin >> item;
    newnode->data = item;
    
    if (head == NULL) {
        newnode->next = NULL;
        head = newnode;
        cout << "Elemento insertado\n";
    } else {
       temp = head;
        while (temp->next != NULL) {
            temp = temp->next;
        }
        temp->next = newnode;
        newnode->next = NULL;
        cout << "Elemento insertado\n";
    }
    }
   
}

void randominsert() {
    struct node *newnode, *temp;
    newnode = (struct node *)malloc(sizeof(struct node));
    int item, pos, i;
    if (newnode == NULL) {
        cout << "OVERFLOW\n";
        
    }else{
         cout << "Ingresa un elemento: ";
        cin >> item;
        newnode->data = item;
        cout << "Ingresa la posicion: ";
        cin >> pos;
    
        temp = head;
        for (i = 0; i < pos -1; i++) {
            if (temp == NULL) {
                cout << "La posicion no existe\n";
                return;
            }
            temp = temp->next;
        }
        newnode->next = temp->next;
        temp->next = newnode;
        cout << "Elemento insertado\n";
    
    }
   
}

void deletebeg() {
    struct node *newNode;
    if (head == NULL) {
        cout << "La lista esta vacia\n";
        
    }else{
         newNode = head;
    head = newNode->next;
    delete newNode;
    cout << "Elemento eliminado desde el principio\n";
    }
   
}

void deleteend() {
    struct node *newNode, *temp;
    if (head == NULL) {
        cout << "La lista esta vacia\n";
        
    }else if(head->next == NULL){
        
     
        delete head;
        head = NULL;
        
        cout << "Solo se elimino un nodo de la lista\n";
        return;
    }
    else{
            newNode = head;
        while (newNode->next != NULL) {
        temp = newNode;
        newNode = newNode->next;
        }
        temp->next = NULL;
        delete newNode;
        cout << "Elemento eliminado desde el final\n";
    
    }
   
   
}

void randomdelete() {
    struct node *newNode, *temp;
    int pos, i;
   
    cout << "Ingresa la posicion: ";
    cin >> pos;
    newNode = head;
    for (i = 0; i < pos; i++) {
        if (newNode == NULL) {
            cout << "La posicion no existe\n";
            return;
        }
        temp = newNode;
        newNode = newNode->next;
    }
    temp->next = newNode->next;
    delete newNode;
    cout << "Elemento eliminado desde la posicion " << pos+1 << "\n";
    
   
}

void display() {
    struct node *newNode;
    newNode = head;
    if (newNode == NULL) {
        cout << "La lista esta vacia\n";
        
    }else{

    cout << "Elementos en la lista: ";
    while (newNode != NULL) {
        cout << newNode->data << " ";
        newNode = newNode->next;
    }
    cout << "\n";
    }
   
}

void search() {
    struct node *newNode;
    int item, pos = 0, flag;
    if (head == NULL) {
        cout << "La lista esta vacia\n";
        
    }else{
         cout << "Ingresa el elemento a buscar: ";
    cin >> item;
    newNode = head;
    while (newNode != NULL) {
        if (newNode->data == item) {
            cout << "Elemento " << item << " encontrado en la posicion " << pos << "\n";
            flag = 0;
            
        }else{
            flag = 1;
        }
        newNode = newNode->next;
        pos++;
    }
    if (flag == 1) {
        cout << "Elemento no encontrado en la lista\n";
    }
    }
   
}

File: Lista Enlazada/Lista.py
class Nodo:
    def __init__(self):
        self.dato = None
        self.next = None


head = None
temp = None


def insertbegin():
    global head
    newNode = Nodo()
    if newNode is None:
        print("Error al crear el nodo")
        return
    else:
        dato = input("Ingrese el dato a agregar al inicio: ")
        newNode.dato = dato
        newNode.next = head
        head = newNode

def insertend():
    global head
    global temp
    newNode = Nodo()
    if newNode is None:
        print("Error al crear el nodo")
    else:
        item = input("Ingrese el dato a agregar al final: ")
        newNode.dato = item
        if head is None:
            newNode.next = None
            head = newNode
            print("El nodo ha sido creado correctamente")
        else:
            temp = head
            while temp.next is not None:
                temp = temp.next
            temp.next = newNode
            newNode.next = None
            print("El nodo ha sido actualizado correctamente")

def randominsert():
    global head
    global temp
    newNode = Nodo()
    if newNode is None:
        print("Error al crear el nodo")
    else:
        item = input("Ingrese el dato a agregar: ")
        newNode.dato = item
        pos= input("Ingrese la posicion a agregar: ")
        temp = head
        i=0
        for i in range(int(pos)-1):
            if temp is None:
                print("Error al agregar el nodo")
                return
            temp = temp.next
        newNode.next = temp.next
        temp.next = newNode
        print("El nodo ha sido actualizado correctamente")

def deletebegin():
    global head
    newNode = Nodo()
    if head is None:
        print("La lista esta vacia")
        return
    else:
        newNode = head
    head =newNode.next
    del newNode
    print("Nodo eliminado correctamente")

def deleteend():
    global head
    global temp
    newNode = Nodo()
    if head is None:
        print("La lista esta vacia")
        return
    else:
        newNode = head
        if newNode.next is None:
            head = None
            del newNode
            print("Nodo eliminado correctamente")
            return
        while newNode.next is not None:
            temp = newNode
            newNode = newNode.next
        temp.next = None
        del newNode
        print("Nodo eliminado correctamente")

def deleterandom():
    global head
    global temp
    newNode = Nodo()
    pos=input("Ingrese la posicion a agregar: ")
    newNode=head
    for i in range(int(pos)-1):
        if newNode is None:
            print("Error al agregar el nodo")
            return
        temp = newNode
        newNode = newNode.next
    temp.next=newNode.next
    del newNode
    print("El nodo eliminado correctamente")

def search():
    newNode = Nodo()
    global head
    pos =0
    if head is None:
        print("La lista esta vacia")
        return
    else:
        item = input("Ingrese el elemento a buscar: ")
        newNode=head
        while newNode is not None:
            if newNode.dato == item:
                print("El nodo se encuentra en la lista")
                flag = 0
                return
            else:
                flag = 1

            newNode = newNode.next
            pos=pos +1
            if flag==1:
                print("El nodo se encuentra en la lista")


def display():
    newNode = Nodo()
    newNode = head
    if newNode is None:
        print("La lista esta vacia")
    else:
        print("Elementos de la lista:")
        while newNode is not None:
            print(newNode.dato + " -> ")
            newNode = newNode.next
        print("\n")

def main():
    choice = 0
    while choice != 9:
        print("1. Agregar al inicio")
        print("2. Agregar al final")
        print("3. Agregar en posicion especifica")
        print("4. Eliminar desde el inicio")
        print("5. Eliminar desde el final")
        print("6. Eliminar desde posicion especifica")
        print("7. Mostrar lista")
        print("8. Buscar")
        print("9. Salir")

        try:
            opcion = input("Seleccione una opcion: ").strip()
            choice = int(opcion)
            match choice:
                case 1:
                    insertbegin()

                case 2:
                    insertend()
                case 3:
                    randominsert()
                case 4:
                    deletebegin()
                case 5:
                    deleteend()
                case 6:
                    deleterandom()
                case 7:
                    display()
                case 8:
                    search()


        except ValueError:
            print("❌ Error: Por favor ingrese un número válido")
if __name__ == "__main__":
    main()